File: tbb/chain/expertseq/randmat.cc
/* randmat: random number generator
 *
 * input:
 *   nrows, ncols: number of rows and columns
 *   s: random number generation seed
 *
 * output:
 *   matrix: random nrows x ncols integer matrix
 */
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>

unsigned char randmat_matrix[20000][20000];

void randmat(int nrows, int ncols, unsigned int s) {
  const int LCG_A = 1664525, LCG_C = 1013904223;
  for (int i = 0; i < nrows; i++) {
    unsigned int seed = s + i;
    for (int j = 0; j < ncols; j++) {
      seed = LCG_A * seed + LCG_C;
      randmat_matrix[i][j] = seed % 100;
    }
  }
}

File: erlang/chain/par/randmat.erl
%
% randmat: random number generation
%
% input:
%   nrows, ncols: the number of rows and columns
%   s: the seed
%
% output:
%   matrix: a nrows x ncols integer matrix
%

-module(randmat).
-export([randmat/3]).
-define(INT_MAX,4294967296).
-define(RAND_MAX,100).
-define(LCG_A,1664525).
-define(LCG_C,1013904223).

randvet(0, _) -> [];
randvet(Ncols, S) ->
  NewS = (?LCG_A * S + ?LCG_C) rem ?INT_MAX,
  [NewS rem ?RAND_MAX | randvet(Ncols - 1, NewS)].

join(Pids) -> [receive {Pid, Result} -> Result end || Pid <- Pids].

randmat(Nrows, Ncols, S) ->
  Parent = self(),
  % parallel_for on rows
  join([spawn(fun() -> Parent ! {self(), randvet(Ncols, S + Row - 1)} end) ||
      Row <- lists:seq(1, Nrows)]).
